### DISCUSSION TITLE
New deployment of IPAM

### DISCUSSION THREAD

**Discussion by JBirSP** [Azure/ipam#238] [Q&A]
Hi!
This looks like a great product that will greatly reduce our admin overhead.
I have a couple of questions that I couldn't really find answer to either in the docs or discussions, maybe I've just missed something. In that case I would appreciate being pointed in the right direction.
1. What is the difference between the WebApp deployment and the Function container deployment? We are mostly going to use the UI, at least initially.
2. I've read some posts in the discussions around private deployment which suits our organisation very well. The documentation on https://azure.github.io/ipam doesn't seem to mention that at all. I did find references that this should be available in version 3.0 of IPAM which seems to have been released in the end of last year.

Thanks!
Joakim

**Comment by DCMattyG** [comment]
Hi @JBirSP , apologies for the delayed response. It's been quite busy lately for me!

Please allow me to respond to your questions:
1. The only real difference is if you prefer to run the solution as a Web App or a Function app. There are a few other small differences, but the primary driver was giving customer choice if they preferred one over the other.
2. The v3.0 version is now a single-container deployment which support Private integration whereas before the multi-container deployment model did not. I just provided a high-level process of creating the Private Endpoints to another users in #244, so perhaps it might be good to review what I shared there to see if that helps.

It's nearly impossible to automatically deploy a Private version of this Application because of the vastly different ways customers have their networking infrastructure deployed in Azure, but it's quite simple to create a few Private Endpoints, Private DNS Zones and to integrate the App Service into a Virtual Network of your choosing.

As always, I'm happy to help out however I can. Please let me know if I can be of further assistance and I'll make myself available for you.

Thanks for the great discussion!

**Reply by JBirSP** [reply]
Hi
Thanks for the answer and instructions (in the other post).
I was successful in deploying the app and it is now using VNet integrations and Private Endpoints. Now on to understanding the concepts of Spaces and Blocks :)

**Reply by DCMattyG** [reply]
That's great to hear @JBirSP!

Here's a quick breakdown of Spaces/Blocks which may help:

**Spaces**
- Spaces are essentially a logical container that only contain non-overlapping IP space
- These are similar to *namespaces* or *IP Spaces* in other products like Infoblox
- You may have multiple *Spaces*, or just a single *Space* depending on your needs
- As a quick example, if you had a *Staging* and *Production* environment where you had the exact same IP addressing because before promoting workload to *Production* you want to make sure everything is working in an identical environment, this would be a use case for two *Spaces* (one for *Staging* and one for *Production*)

**Blocks**
- These are like CIDR ranges that are assigned to either a workload or line of business that will then be divided into Virtual Networks
- For example, perhaps your I.T. team and Research team need to build resources in Azure and want some autonomy. You could create a large *Block* for the I.T. team and another large *Block* for the Research team and they could carve that into smaller CIDR's which in turn get assigned to new Virtual Networks in Azure.
- You can also assign existing Virtual Networks to *Blocks* for brownfield type scenarios to properly track CIDR utilization

Hope that brief overview helps, let me know if you need any further clarification on the above!

**Reply by JBirSP** [reply]
Thank you for those details. 
For our environment I added a single space as we have no overlapping IP ranges (other than for isolated instances of Kubernetes etc. but they are not peered and do not need tracking). I also set each block to equal one existing Vnet as we did not plan for super netting IP ranges for different departments or business areas.
I find many of the features very useful and we have started using it to find available subnets.
I may have configured something wrong though. In the Discover -> Spaces section it marks our space as 100% utilized, and moving to the Blocks section, they also show up as 100% utilized even though the Vnets still have room for more subnets.

I've found the planner very useful to set up new subnets within a Vnet, and I would also like to have something similar on the "space" level to help set up new Vnet (or rather discover available IP address space that is not in use by any existing Vnet).

Maybe I've done some strange configurations, but I think we might get more out of this when we've learned more.

**Reply by DCMattyG** [reply]
Hi there @JBirSP, always happy to have feedback on how we can make things better!

Quick note on the Space/Block utilization, at this point in time Block utilization is reflective of how much unallocated Block space there is (e.g. how much of a Block's CIDR is **not** allocated for Virtual Networks), and similarly Space utilization is a roll-up of this. So the short answer is, at this level it's not a calculation down to the Subnet level.

For your other items, perhaps we can setup a few minutes for us to chat on Team about any improvements you'd like to see in the product or perhaps to answer some how-to questions live? If so, please drop me an email at Matthew.Garrett@microsoft.com and we can figure out a time that works best for you.

Much appreciated!
